server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java

```diff
@@ -246,11 +246,12 @@ protected void doRun() throws Exception {
             Translog.Operation operation;
             while ((operation = snapshot.next()) != null) {
                 final long seqNo = operation.seqNo();
-                if (startingSeqNo >= 0 &&
-                    (seqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || seqNo < startingSeqNo)) {
+                // do not send operations if the global checkpoint is still unassigned.
+                if (seqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || startingSeqNo < 0 || seqNo < startingSeqNo) {
```

Comment:
I wonder if this is equivalent to the weaker `if (seqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || seqNo < startingSeqNo) {`, i.e., we could just remove `startingSeqNo >= 0` from the condition.

If the global checkpoint is `SequenceNumbers.UNASSIGNED_SEQ_NO`, I would assume that most operations (except a few that might have happened before global checkpoint has advanced) would be `SequenceNumbers.UNASSIGNED_SEQ_NO`.